packages/ui/src/elements/Upload/index.tsx

```diff
@@ -194,42 +192,52 @@ export const Upload: React.FC<UploadProps> = (props) => {
-            const response = await fetch(`${serverURL}${api}${pasteURL}`)
-
-            if (!response.ok) {
-              const errorData = await response.json()
-              throw new Error(errorData.error || 'Failed to fetch the file from the server')
-            }
-
-            const blob = await response.blob()
-            const fileName = fileUrl.split('/').pop()
-            const file = new File([blob], fileName, { type: blob.type })
-
-            handleFileChange(file)
-            setUploadStatus('idle')
-          } else {
-            throw new Error('The provided URL is not allowed.')
+        // Attempt client-side fetch first
+        const response = await fetch(fileUrl)
+
+        if (!response.ok) {
+          throw new Error(`Client-side fetch failed with status: ${response.status}`)
+        }
+
+        const data = await response.blob()
+        const fileName = decodeURIComponent(fileUrl.split('/').pop() || '')
+
+        const file = new File([data], fileName, { type: data.type })
+        handleFileChange(file)
+        setUploadStatus('idle')
+        return // Exit if client-side fetch succeeds
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      } catch (error) {
+        // Continue to server-side fallback
+      }
+
+      // If pasteURL.allowList is defined, try to server-side fetch
+      if (useServerSideFetch) {
+        try {
+          const pasteURL = `/${collectionSlug}/paste-url${id ? `/${id}?` : '?'}src=${encodeURIComponent(fileUrl)}`
+          const response = await fetch(`${serverURL}${api}${pasteURL}`)
+
+          if (!response.ok) {
+            const errorData = await response.json()
+            throw new Error(errorData.error || 'Server-side fetch failed.')
           }
-        } else {
-          // Perform client-side fetch if no allowList is specified
-          const response = await fetch(fileUrl)
-          const data = await response.blob()
 
-          // Extract the file name from the URL
-          const fileName = fileUrl.split('/').pop()
+          const blob = await response.blob()
+          const fileName = decodeURIComponent(fileUrl.split('/').pop() || '')
+          const file = new File([blob], fileName, { type: blob.type })
 
-          // Create a new File object from the Blob data
-          const file = new File([data], fileName, { type: data.type })
           handleFileChange(file)
           setUploadStatus('idle')
+          return // Exit if server-side fetch succeeds
+        } catch (serverError) {
+          toast.error(`${serverError.message} The provided URL is not allowed.`)
+          setUploadStatus('failed')
         }
-      } catch (e) {
-        toast.error(e.message)
+      } else {
+        // Only show error if all attempts have failed
+        toast.error(
+          'Client-side fetch is disabled or failed, and no server-side fallback is available.',
```

Comment:
This is too much detail for the end users and isn't translated.
It might be more simple to understand what the error outcome is if you make a variable for errorToast at the top, assign it at L211 and L222, make early returns if there is success and then only call the toast at end. With the complex if/else/try/catch structure it is hard to follow.